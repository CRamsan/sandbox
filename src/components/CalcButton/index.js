import React from 'react'
import PropTypes from 'prop-types'
import { Entity } from 'aframe-react'
import { Text } from 'components'

export default class CalcButton extends React.Component {
  static propTypes = {
    actionToTrigger: PropTypes.func,
    id: PropTypes.string,
    buttonColor: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    position: PropTypes.object,
    text: PropTypes.string,
    textColor: PropTypes.string,
    textWidth: PropTypes.number,
    value: PropTypes.string,
  }

  static defaultProps = {
    value: '',
    width: 0.15,
    height: 0.15,
    position: { x: 0, y: 0, z: 0 },
    text: 'Text',
    textColor: '#000',
    buttonColor: '#bada55',
    textWidth: 0.8,
  }

  constructor(props) {
    super(props)

    this.state = {
      depth: 0.02,
      opacity: 1,
    }

    this.startIntersection = this.startIntersection.bind(this)
    this.endIntersection = this.endIntersection.bind(this)
  }

  startIntersection() {
    const { actionToTrigger, value } = this.props
    this.setState({ depth: 0, opacity: 0.2 })
    actionToTrigger(value)
  }

  endIntersection() {
    this.setState({ depth: 0.02, opacity: 1 })
  }

  render() {
    const { id, position, width, height, buttonColor, text, textColor, textWidth } = this.props
    const { depth, opacity } = this.state
    return (
      <Entity
        id={id}
        className="interactive"
        geometry={{ primitive: 'box', height, width, depth }}
        material={{ shader: 'flat', side: 'double', color: buttonColor, opacity }}
        scale={{ x: 0.5, y: 0.5, z: 0.5 }}
        position={position}
        hoverable
        clickable
        events={{
          'hover-start': this.startIntersection,
          'hover-end': this.endIntersection,
        }}
      >
        <Text
          value={text}
          color={textColor}
          align="center"
          zOffset={0.02}
          width={textWidth}
        />
      </Entity>
    )
  }
}
