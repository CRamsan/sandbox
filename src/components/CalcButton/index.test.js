import React from 'react'
import renderer from 'react-test-renderer'
import CalcButton from '.'

jest.mock('react-dom')

describe('CalcButton', () => {
  it('renders <a-entity>', () => {
    const tree = renderer.create(<CalcButton />).toJSON()
    expect(tree.type).toBe('a-entity')
  })

  it('renders <a-entity> with id', () => {
    const tree = renderer.create(
      <CalcButton id="calc-button" />
    ).toJSON()
    expect(tree.props.id).toBe('calc-button')
  })

  it('is interactable with super-hands', () => {
    const tree = renderer.create(
      <CalcButton />
    ).toJSON()
    expect(tree.props.className).toBe('interactive')
  })

  it('renders the entity consistently', () => {
    const action = jest.fn()
    const tree = renderer.create(<CalcButton id="calc-button" actionToTrigger={action} value="test" />)

    expect(tree).toMatchSnapshot()
  })
})
