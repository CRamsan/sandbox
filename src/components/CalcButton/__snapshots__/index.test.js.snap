// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalcButton renders the entity consistently 1`] = `
<a-entity
  className="interactive"
  id="calc-button"
>
  <span />
</a-entity>
`;
